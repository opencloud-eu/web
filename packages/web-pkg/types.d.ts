import { Ability } from '@opencloud-eu/web-client'
import {
  ArchiverService,
  ClientService,
  LoadingService,
  PreviewService,
  PasswordPolicyService
} from './src/services'
import { Route, Router } from 'vue-router'

export * from './src'

declare module 'vue' {
  interface ComponentCustomProperties {
    $ability: Ability
    $archiverService: ArchiverService
    $clientService: ClientService
    $loadingService: LoadingService
    $previewService: PreviewService
    $passwordPolicyService: PasswordPolicyService
    $router: Router
    $route: Route
  }
}
