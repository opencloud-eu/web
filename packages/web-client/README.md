# web-client

The `web-client` is a standalone package that allows you to interact with the [OpenCloud](https://github.com/opencloud-eu/opencloud/) APIs via TypeScript. It provides an abstraction layer between the server and a (web-) application that converts API data into objects with helpful types and utilities. This abstraction ensures that users of the APIs don't need in-depth knowledge about them, such as required methods or returned status codes.

The supported APIs are:

- Graph (drives, sharing, user & group management)
- OCS (capabilities & url signing)
- WebDAV (file operations)

## Installation

Depending on your package manager, run one of the following commands:

```
$ npm install @opencloud-eu/web-client

$ pnpm add @opencloud-eu/web-client

$ yarn add @opencloud-eu/web-client
```

## Usage

### Graph

The graph client needs to be instantiated with a base URI corresponding to your OpenCloud deployment and an axios instance. The axios instance is being used for all requests, which means it needs to include all relevant headers either statically or via interceptor.

```
import axios from axios
import { graph } from '@opencloud-eu/web-client'

const accessToken = 'some_access_token'
const baseURI = 'some_base_uri'

const axiosClient = axios.create({
	headers: { Authorization: accessToken }
})

const graphClient = graph(baseURI, axiosClient)
```

The following example demonstrates how to retrieve all spaces accessible to the user. A `SpaceResource` can then be used to e.g. fetch files and folders (see webdav example down below).

```
const mySpaces = await graphClient.drives.listMyDrives()
```

### OCS

The ocs client needs to be instantiated with a base URI corresponding to your OpenCloud deployment and an axios instance. The axios instance is being used for all requests, which means it needs to include all relevant headers either statically or via interceptor.

```
import axios from axios
import { ocs } from '@opencloud-eu/web-client'

const accessToken = 'some_access_token'
const baseURI = 'some_base_uri'

const axiosClient = axios.create({
	headers: { Authorization: accessToken }
})

const ocsClient = ocs(baseURI, axiosClient)
```

The following examples demonstrate how to fetch capabilities and sign URLs.

```
const capabilities = await ocsClient.getCapabilities()

const signedUrl = await ocsClient.signUrl('some_url_to_sign', 'your_username')
```

### WebDav

The webdav client needs to be instantiated with a base URI corresponding to your OpenCloud deployment. You can also pass a header callback which will be called with every dav request.

```
import { webdav } from '@opencloud-eu/web-client'

const accessToken = 'some_access_token'
const baseURI = 'some_base_uri'

const webDavClient = webdav(
	baseURI,
	() => ({ Authorization: accessToken })
)
```

The following example demonstrates how to list all resources of a given `SpaceResource` (see above how to fetch space resources).

```
const { resource, children } = await webDavClient.listFiles(spaceResource)
```
