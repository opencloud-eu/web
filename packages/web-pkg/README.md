# web-pkg

`web-pkg` is a package that provides utilities, most importantly a variety of components and composables, that can be useful when developing apps and extensions for OpenCloud Web.

## Installation

Depending on your package manager, run one of the following commands:

```
$ npm install @opencloud-eu/web-pkg --save-dev

$ pnpm add -D @opencloud-eu/web-pkg

$ yarn add @opencloud-eu/web-pkg --dev
```

It's recommended to install this package as a dev dependency because it's only really needed for providing autocompletion in your IDE and unit tests. In a runtime context, the OpenCloud Web runtime provides the actual implementation.
