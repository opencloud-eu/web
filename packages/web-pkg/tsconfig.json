{
  "extends": "@opencloud-eu/tsconfig",

  // necessary for building web-pkg standalone, otherwise we get errors with inferred types
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "@vue/shared": ["node_modules/@vue/shared/"],
      "@codemirror/autocomplete": ["./node_modules/@codemirror/autocomplete"],
      "vue/jsx-runtime": ["./node_modules/vue/jsx-runtime"],
      "@popperjs/core": ["./../design-system/node_modules/@popperjs/core"]
    }
  },

  // include component types of the design-system
  "include": ["../design-system/types.d.ts"]
}
