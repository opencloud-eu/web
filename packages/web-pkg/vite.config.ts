import { join, resolve } from 'path'
import { defineConfig, searchForWorkspaceRoot } from 'vite'
import dts from 'vite-plugin-dts'
import { nodePolyfills } from 'vite-plugin-node-polyfills'
import vue from '@vitejs/plugin-vue'
import pkg from './package.json'

const projectRootDir = searchForWorkspaceRoot(process.cwd())
const external = [...Object.keys(pkg.dependencies)]

export default defineConfig({
  resolve: {
    alias: {
      crypto: join(projectRootDir, 'polyfills/crypto.js')
    }
  },
  css: {
    preprocessorOptions: {
      scss: {
        additionalData: `
          @use "sass:math";
          @use "sass:string";
          @use "sass:meta";
        `,
        silenceDeprecations: ['legacy-js-api', 'import']
      }
    }
  },
  build: {
    lib: {
      entry: resolve(__dirname, 'src/index.ts'),
      name: 'web-pkg',
      fileName: 'web-pkg'
    },
    rollupOptions: {
      external: external.filter(
        (e) =>
          // something is off with this lib, see https://github.com/ahmadjoya/generate-password-lite/issues/8
          e !== 'js-generate-password'
      )
    }
  },
  plugins: [
    vue(),
    nodePolyfills({
      exclude: ['crypto']
    }),
    dts({ exclude: ['**/tests'] })
  ]
})
