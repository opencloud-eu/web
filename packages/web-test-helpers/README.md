# web-test-helpers

This packages provides utilities for unit testing within the OpenCloud app ecosystem.

## Installation

Depending on your package manager, run one of the following commands:

```
$ npm install @opencloud-eu/web-test-helpers --save-dev

$ pnpm add -D @opencloud-eu/web-test-helpers

$ yarn add @opencloud-eu/web-test-helpers --dev
```

Make sure that you have `vue`, `@vue/test-utils`, and `@opencloud-eu/web-pkg` as dev dependencies of your app since those are peer dependencies of this package.
