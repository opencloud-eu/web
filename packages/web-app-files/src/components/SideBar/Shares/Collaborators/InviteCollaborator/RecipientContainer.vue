<template>
  <oc-recipient
    :data-testid="`recipient-container-${formattedRecipient.name}`"
    class="files-share-invite-recipient"
    :recipient="formattedRecipient"
  >
    <template #avatar>
      <user-avatar
        v-if="recipient.shareType === ShareTypes.user.value"
        :user-id="recipient.id"
        :user-name="recipient.displayName"
        width="16.8"
      />
    </template>
    <template #append>
      <oc-button
        class="files-share-invite-recipient-btn-remove raw-hover-surface"
        appearance="raw"
        :aria-label="btnDeselectRecipientLabel"
        @click.stop="deselect(recipient)"
      >
        <oc-icon name="close" size="small" />
      </oc-button>
    </template>
  </oc-recipient>
</template>

<script lang="ts">
import { CollaboratorAutoCompleteItem, ShareTypes } from '@opencloud-eu/web-client'
import { computed, defineComponent, PropType } from 'vue'
import { Recipient } from '@opencloud-eu/design-system/helpers'
import { UserAvatar } from '@opencloud-eu/web-pkg'

export default defineComponent({
  components: { UserAvatar },
  props: {
    recipient: {
      type: Object as PropType<CollaboratorAutoCompleteItem>,
      required: true
    },
    deselect: {
      type: Function,
      required: false,
      default: null
    }
  },
  setup(props) {
    const externalIssuer = computed(() => {
      if (props.recipient.shareType === ShareTypes.remote.value) {
        return props.recipient.identities?.[0]?.issuer
      }
      return ''
    })

    return {
      externalIssuer,
      ShareTypes
    }
  },
  data(): { formattedRecipient: Recipient } {
    let name = this.recipient.displayName
    if (this.externalIssuer) {
      name += ` (${this.externalIssuer})`
    }

    return {
      formattedRecipient: {
        name,
        icon: this.getRecipientIcon()
      }
    }
  },

  computed: {
    btnDeselectRecipientLabel() {
      return this.$gettext('Deselect %{name}', { name: this.recipient.displayName })
    }
  },

  methods: {
    getRecipientIcon() {
      switch (this.recipient.shareType) {
        case ShareTypes.group.value:
          return {
            name: ShareTypes.group.icon,
            label: this.$gettext('Group')
          }

        case ShareTypes.guest.value:
          return {
            name: ShareTypes.guest.icon,
            label: this.$gettext('Guest user')
          }

        case ShareTypes.remote.value:
          return {
            name: ShareTypes.remote.icon,
            label: this.$gettext('External user')
          }

        default:
          return {
            name: ShareTypes.user.icon,
            label: this.$gettext('User')
          }
      }
    }
  }
})
</script>

<style lang="scss">
.files-share-invite-recipient {
  margin: 4px 2px 0;
  padding: 0 0.25em;
  overflow-wrap: anywhere;

  .oc-recipient-avatar {
    min-width: 16.8px;
  }
}
</style>
