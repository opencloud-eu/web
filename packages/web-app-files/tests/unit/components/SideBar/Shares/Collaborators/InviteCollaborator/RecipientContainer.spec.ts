import RecipientContainer from '../../../../../../../src/components/SideBar/Shares/Collaborators/InviteCollaborator/RecipientContainer.vue'
import { CollaboratorAutoCompleteItem, ShareTypes } from '@opencloud-eu/web-client'
import { defaultPlugins, mount } from '@opencloud-eu/web-test-helpers'
import { CapabilityStore } from '@opencloud-eu/web-pkg'

vi.mock('../../../../../../../src/helpers/user/avatarUrl', () => ({
  avatarUrl: vi.fn().mockReturnValue('avatarUrl')
}))

const getRecipient = (shareType: number = ShareTypes.user.value): CollaboratorAutoCompleteItem => ({
  displayName: 'Albert Einstein',
  id: 'einstein',
  shareType
})

describe('InviteCollaborator RecipientContainer', () => {
  describe('renders a recipient with a deselect button', () => {
    it.each(ShareTypes.authenticated)('different recipients for different shareTypes', (type) => {
      const recipient = getRecipient(type.value)
      const { wrapper } = getMountedWrapper(recipient)
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
  it('displays an avatar image if capability is present', async () => {
    const recipient = getRecipient()
    const { wrapper } = getMountedWrapper(recipient, true)
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('emits an event if deselect button is clicked', async () => {
    const recipient = getRecipient()
    const { wrapper } = getMountedWrapper(recipient, true)
    const spyOnDeselect = (wrapper.vm.deselect as any).mockImplementation()
    const button = wrapper.find('.files-share-invite-recipient-btn-remove')
    await button.trigger('click')
    expect(spyOnDeselect).toHaveBeenCalledTimes(1)
  })
})

function getMountedWrapper(recipient: CollaboratorAutoCompleteItem, avatarsEnabled = false) {
  const capabilities = {
    files_sharing: {
      user: {
        profile_picture: avatarsEnabled
      }
    }
  } satisfies Partial<CapabilityStore['capabilities']>

  return {
    wrapper: mount(RecipientContainer, {
      props: {
        recipient,
        deselect: vi.fn()
      },
      global: {
        plugins: [...defaultPlugins({ piniaOptions: { capabilityState: { capabilities } } })],
        stubs: {
          OcIcon: true
        }
      }
    })
  }
}
