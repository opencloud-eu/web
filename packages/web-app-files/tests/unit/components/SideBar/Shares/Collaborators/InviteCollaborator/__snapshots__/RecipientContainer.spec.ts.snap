// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InviteCollaborator RecipientContainer > displays an avatar image if capability is present 1`] = `
"<span class="oc-recipient files-share-invite-recipient" data-testid="recipient-container-Albert Einstein"><span class="vue-avatar--wrapper oc-avatar oc-recipient-avatar" style="width: 16.8px; height: 16.8px; line-height: 16.8px; font-size: 6px; font-family: Helvetica, Arial, sans-serif; color: white;" width="16.8" aria-hidden="true" focusable="false" data-test-user-name="Albert Einstein" data-testid="recipient-avatar"><img src="avatarUrl" alt="" aria-hidden="true" loading="lazy" class="avatarImg"></span>
<p class="oc-recipient-name" data-testid="recipient-name">Albert Einstein</p> <!-- @slot Append content (actions, additional info, etc.)  --> <button type="button" aria-label="Deselect Albert Einstein" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-secondary oc-button-raw oc-button-secondary-raw files-share-invite-recipient-btn-remove raw-hover-surface">
  <!--v-if-->
  <!-- @slot Content of the button -->
  <oc-icon-stub accessiblelabel="" color="" filltype="fill" name="close" size="small" type="span"></oc-icon-stub>
</button></span>"
`;

exports[`InviteCollaborator RecipientContainer > renders a recipient with a deselect button > different recipients for different shareTypes 1`] = `
"<span class="oc-recipient files-share-invite-recipient" data-testid="recipient-container-Albert Einstein"><span class="vue-avatar--wrapper oc-avatar oc-recipient-avatar" style="width: 16.8px; height: 16.8px; line-height: 16.8px; background-color: #c21c53; font-size: 6px; font-family: Helvetica, Arial, sans-serif; color: white;" width="16.8" aria-hidden="true" focusable="false" data-test-user-name="Albert Einstein" data-testid="recipient-avatar"><span class="avatar-initials" style="color: white;">AE</span></span>
<p class="oc-recipient-name" data-testid="recipient-name">Albert Einstein</p> <!-- @slot Append content (actions, additional info, etc.)  --> <button type="button" aria-label="Deselect Albert Einstein" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-secondary oc-button-raw oc-button-secondary-raw files-share-invite-recipient-btn-remove raw-hover-surface">
  <!--v-if-->
  <!-- @slot Content of the button -->
  <oc-icon-stub accessiblelabel="" color="" filltype="fill" name="close" size="small" type="span"></oc-icon-stub>
</button></span>"
`;

exports[`InviteCollaborator RecipientContainer > renders a recipient with a deselect button > different recipients for different shareTypes 2`] = `"<span class="oc-recipient files-share-invite-recipient" data-testid="recipient-container-Albert Einstein"><oc-icon-stub accessiblelabel="Group" color="" filltype="fill" name="group" size="small" type="span" class="oc-recipient-icon" data-testid="recipient-icon"></oc-icon-stub> <p class="oc-recipient-name" data-testid="recipient-name">Albert Einstein</p> <!-- @slot Append content (actions, additional info, etc.)  --> <button type="button" aria-label="Deselect Albert Einstein" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-secondary oc-button-raw oc-button-secondary-raw files-share-invite-recipient-btn-remove raw-hover-surface"><!--v-if--> <!-- @slot Content of the button --> <oc-icon-stub accessiblelabel="" color="" filltype="fill" name="close" size="small" type="span"></oc-icon-stub></button></span>"`;

exports[`InviteCollaborator RecipientContainer > renders a recipient with a deselect button > different recipients for different shareTypes 3`] = `"<span class="oc-recipient files-share-invite-recipient" data-testid="recipient-container-Albert Einstein"><oc-icon-stub accessiblelabel="Guest user" color="" filltype="fill" name="global" size="small" type="span" class="oc-recipient-icon" data-testid="recipient-icon"></oc-icon-stub> <p class="oc-recipient-name" data-testid="recipient-name">Albert Einstein</p> <!-- @slot Append content (actions, additional info, etc.)  --> <button type="button" aria-label="Deselect Albert Einstein" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-secondary oc-button-raw oc-button-secondary-raw files-share-invite-recipient-btn-remove raw-hover-surface"><!--v-if--> <!-- @slot Content of the button --> <oc-icon-stub accessiblelabel="" color="" filltype="fill" name="close" size="small" type="span"></oc-icon-stub></button></span>"`;

exports[`InviteCollaborator RecipientContainer > renders a recipient with a deselect button > different recipients for different shareTypes 4`] = `"<span class="oc-recipient files-share-invite-recipient" data-testid="recipient-container-Albert Einstein"><oc-icon-stub accessiblelabel="External user" color="" filltype="fill" name="earth" size="small" type="span" class="oc-recipient-icon" data-testid="recipient-icon"></oc-icon-stub> <p class="oc-recipient-name" data-testid="recipient-name">Albert Einstein</p> <!-- @slot Append content (actions, additional info, etc.)  --> <button type="button" aria-label="Deselect Albert Einstein" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-secondary oc-button-raw oc-button-secondary-raw files-share-invite-recipient-btn-remove raw-hover-surface"><!--v-if--> <!-- @slot Content of the button --> <oc-icon-stub accessiblelabel="" color="" filltype="fill" name="close" size="small" type="span"></oc-icon-stub></button></span>"`;
